asynciterator.ts

```diff
@@ -1526,101 +1529,88 @@ export class UnionIterator<T> extends BufferedIterator<T> {
-        this._fillBuffer();
-      });
-    }
+    else
+      this.readable = true;
   }
 
   // Adds the given source to the internal sources array
-  protected _addSource(source: MaybePromise<InternalSource<T>>) {
-    if (isPromise(source))
-      source = wrap<T>(source) as any as InternalSource<T>;
-    if (!source.done) {
-      this._sources.push(source);
-      source._destination = this;
-      source.on('error', destinationEmitError);
-      source.on('readable', destinationFillBuffer);
-      source.on('end', destinationRemoveEmptySources);
-    }
+  protected _addSource(source: InternalSource<T>) {
+    source._destination = this;
+    source.on('error', destinationEmitError);
+    source.on('readable', destinationSetReadable);
+    source.on('end', destinationRemoveEmptySources);
+  }
+
+  protected _removeSource(source: InternalSource<T>) {
+    source.removeListener('error', destinationEmitError);
+    source.removeListener('readable', destinationSetReadable);
+    source.removeListener('end', destinationRemoveEmptySources);
+    // @ts-ignore
+    delete source._destination;
   }
 
-  // Removes sources that will no longer emit items
   protected _removeEmptySources() {
-    this._sources = this._sources.filter((source, index) => {
-      // Adjust the index of the current source if needed
-      if (source.done && index <= this._currentSource)
-        this._currentSource--;
-      return !source.done;
+    this.buffer.filter((source: any) => {
+      if (source.done) {
+        this._removeSource(source);
+        return false;
+      }
+      return true;
     });
-    this._fillBuffer();
+    if (this.buffer.empty && this._sources.done && this._sourceStarted)
+      this.close();
+  }
+
+  public close() {
+    this._removeSource(this._sources as any);
+    super.close();
   }
 
+
   // Reads items from the next sources
-  protected _read(count: number, done: () => void): void {
-    // Start source loading if needed
-    if (this._pending?.sources)
-      this._loadSources();
-
-    // Try to read `count` items
-    let lastCount = 0, item : T | null;
-    while (lastCount !== (lastCount = count)) {
-      // Try every source at least once
-      for (let i = 0; i < this._sources.length && count > 0; i++) {
-        // Pick the next source
-        this._currentSource = (this._currentSource + 1) % this._sources.length;
-        const source = this._sources[this._currentSource];
-        // Attempt to read an item from that source
-        if ((item = source.read()) !== null) {
-          count--;
-          this._push(item);
-        }
-      }
+  public read(): T | null {
+    if (!this._sourceStarted)
+      this._sourceStarted = true;
+
+    const { buffer, _sources } = this;
+    let item: T | null;
+    let iterator: AsyncIterator<T> | null;
+    for (iterator of buffer) {
+      if (iterator.readable && (item = iterator.read()) !== null)
+        return item;
     }
+    while ((iterator = _sources.read()) !== null) {
```

Comment:
If I understand correctly, we get here when all of the sources that we are currently reading from fail to return an item. If that happens, we add as many new sources as we can.

Can we add a limit to the number of sources we can be reading from at any one time? That would come useful in use cases where a source is backed by a stateful form of I/O with allocated resources (i.e. never good to have too many open connections or file descriptors).